src/x/io/compression.go

```diff
@@ -0,0 +1,143 @@
+// Copyright (c) 2020 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package io
+
+import (
+	"bufio"
+	"bytes"
+	"fmt"
+	"io"
+
+	snappy "github.com/golang/snappy"
+)
+
+// The snappy compression stream identifier. A valid snappy framed stream always
+// starts with this sequence of bytes.
+// https://github.com/google/snappy/blob/master/framing_format.txt#L68
+var snappyStreamID = []byte{0xff, 0x06, 0x00, 0x00, 0x73, 0x4e, 0x61, 0x50, 0x70, 0x59}
```

Comment:
Instead of relying on snappy-specific magic value, given m3msg protocol is basically 64bit varint-delimited opaque payloads, could we just reserve a magic value, eg. the negative range, to indicate compression method instead? I'm also looking into adding zstd/snappy support into m3aggregator clients and it would be nice to have a future-proof way of determining encoding.
cc @mway

Reply:
Yea. This sounds like it would make sense. I didn't notice actually that the msg length prefix has room for additional magic values.